python/ray/data/_internal/execution/operators/actor_pool_map_operator.py

```diff
@@ -0,0 +1,356 @@
+        self._dispatch_tasks()
+
+    def _dispatch_tasks(self):
+        """Try to dispatch tasks from the bundle buffer to the actor pool.
+
+        This is called when:
+            * a new input bundle is added,
+            * a task finishes,
+            * a new worker has been created.
+        """
+        while self._bundle_queue:
+            # Pick an actor from the pool.
+            actor = self._actor_pool.pick_actor()
+            if actor is None:
+                # No actors available for executing the next task.
+                break
+            # Submit the map task.
+            bundle = self._bundle_queue.popleft()
+            input_blocks = [block for block, _ in bundle.blocks]
+            ref = actor.submit.options(num_returns="dynamic").remote(
+                self._transform_fn_ref, *input_blocks
+            )
+            task = _TaskState(bundle)
+            self._tasks[ref] = (task, actor)
+            self._handle_task_submitted(task)
+
+    def notify_work_completed(
+        self, ref: Union[ObjectRef[ObjectRefGenerator], ray.ObjectRef]
+    ):
+        # This actor pool MapOperator implementation has both task output futures AND
+        # worker started futures to handle here.
+        if ref in self._tasks:
+            # Get task state and set output.
+            task, actor = self._tasks.pop(ref)
+            task.output = self._map_ref_to_ref_bundle(ref)
+            self._handle_task_done(task)
+            # Return the actor that was running the task to the pool.
+            self._actor_pool.return_actor(actor)
+        else:
+            # ref is a future for a now-ready actor; move actor from pending to the
+            # active actor pool.
+            self._actor_pool.pending_to_running(ref)
+            # TODO(Clark): Add idle timeouts?
+        # Try to dispatch queued tasks.
+        self._dispatch_tasks()
+
+    def inputs_done(self):
+        # Call base implementation to handle any leftover bundles.
+        super().inputs_done()
+
+        self._inputs_done = True
+        if not self._bundle_queue:
+            # No more tasks will be submitted, so we kill all current and future
+            # inactive workers.
+            self._actor_pool.drain()
+
+    def shutdown(self):
+        # We kill all actors in the pool on shutdown, even if they are busy doing work.
+        self._actor_pool.kill_all_actors()
+        super().shutdown()
+
+    def get_work_refs(self) -> List[ray.ObjectRef]:
+        # Work references that we wish the executor to wait on includes both task
+        # futures AND worker ready futures.
+        return list(self._tasks.keys()) + self._actor_pool.get_pending_actor_refs()
+
+    def num_active_work_refs(self) -> int:
+        # Active work references only includes running tasks, not pending actor starts.
+        return len(self._tasks)
+
+    def progress_str(self) -> str:
+        return f"{self._actor_pool.num_total_actors()}/" f"{self._pool_size} actors"
```

Comment:
For a static pool, this will always show `N/N`; I made this change to bring the progress string closer in line with the legacy actor compute strategy.

Autoscaling isn't a concern here since this won't be the implementation in that PR; this is changed in the autoscaling PR to
```python
        return (
            f"{self._actor_pool.num_total_actors()}/"
            f"{self._autoscaling_policy.max_workers} actors"
        )
```

Reply:
Updated to the format that you indicated in your other comment!